apple/testing/default_runner/macos_test_runner.template.sh

```diff
@@ -160,13 +166,41 @@ xcrun llvm-cov \
   "$test_binary" \
   @"$COVERAGE_MANIFEST" \
   > "$COVERAGE_OUTPUT_FILE" \
-  2> "$error_file" \
-  || llvm_cov_status=$?
+  2> "$export_error_file" \
+  || llvm_cov_export_status=$?
+
+llvm_cov_json_export_status=0
+readonly json_export_error_file="$TEST_TMP_DIR/llvm-cov-json-export-error.txt"
+if [ "$COVERAGE_PRODUCE_JSON" = true ]; then
+  if [[ -n "${TEST_UNDECLARED_OUTPUTS_DIR}" ]]; then
+    OUTPUT_DIR="${TEST_UNDECLARED_OUTPUTS_DIR}"
+    mkdir -p "$OUTPUT_DIR"
+    xcrun llvm-cov \
+      export \
+      -format text \
+      -instr-profile "$profdata" \
+      -ignore-filename-regex='.*external/.+' \
+      -path-equivalence="$ROOT",. \
+      "$test_binary" \
+      @"$COVERAGE_MANIFEST" \
+      > "$TEST_UNDECLARED_OUTPUTS_DIR/coverage.json"
+        2> "$json_export_error_file" \
+        || llvm_cov_json_export_status=$?
+  fi
+fi
 
 # Error ourselves if lcov outputs warnings, such as if we misconfigure
 # something and the file path of one of the covered files doesn't exist
-if [[ -s "$error_file" || "$llvm_cov_status" -ne 0 ]]; then
+if [[ -s "$export_error_file" || "$llvm_cov_export_status" -ne 0 ]]; then
   echo "error: while exporting coverage report" >&2
-  cat "$error_file" >&2
+  cat "$export_error_file" >&2
   exit 1
 fi
+
+if [ "$COVERAGE_PRODUCE_JSON" = true ]; then
+  if [[ -s "$json_export_error_file" || "$llvm_cov_json_export_status" -ne 0 ]]; then
```

Comment:
once you move the other block up you can add this inside the if that actually produces it so we don't have to dup the conditional